File: scala/janken-for-couple/src/main/scala/player/Strategy.scala
package philomagi.practices_by_game.janken_for_couple
package player

import hand.Hand

trait Strategy {
  def select: Hand
}
object Strategy {
  object Only {
    object OnlyStone extends Strategy {
      override def select: Hand = Hand.Stone
    }
    object OnlyScissors extends Strategy {
      override def select: Hand = Hand.Scissors
    }
    object OnlyPaper extends Strategy {
      override def select: Hand = Hand.Paper
    }
  }
}

File: haskell/janken-couple/src/JankenCouple/Strategy.hs
module JankenCouple.Strategy where

import JankenCouple.Hand

data Strategy =
  OnlyStone |
  OnlyScissors |
  OnlyPaper |
  OrderedStrategy { current :: Hand }
  deriving (Eq, Show)

orderedFrom :: Hand -> Strategy
orderedFrom h = OrderedStrategy h

select :: Strategy -> Hand
select OnlyStone = Stone
select OnlyScissors = Scissors
select OnlyPaper = Paper
select (OrderedStrategy current) = current

next :: Strategy -> Strategy
next (OrderedStrategy Stone) = OrderedStrategy(Scissors)
next (OrderedStrategy Scissors) = OrderedStrategy(Paper)
next (OrderedStrategy Paper) = OrderedStrategy(Stone)
next s = s
